lib/compiler.js

```diff
@@ -606,16 +607,37 @@ function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
   return (fixedFile === undefined ? resolvedPath : fixedFile);
 }
 
-function buildAllMixinInstructions(appRootDir, mixinDirs, options) {
+function buildAllMixinInstructions(appRootDir, mixinDirs, mixinSources, options,
+  modelInstructions) {
   var extensions = _.without(_.keys(require.extensions),
     _.keys(getExcludedExtensions()));
   var files = options.mixins || [];
 
-  mixinDirs.forEach(function(dir) {
+  var mixins = loadMixins(appRootDir, mixinDirs, files, extensions, options);
+  if (mixins === undefined) return;
+
+  // Fetch unique list of mixin names, used in models
+  var modelMixins = _.unique(fetchMixinsUsedInModels(modelInstructions));
+
+  files = [];
+  var mixinsFromMixinSources = loadMixins(appRootDir, mixinSources, files,
+    extensions, options);
+  if (mixinsFromMixinSources === undefined) return;
+
+  // Filter in only mixins, that are used in models
+  mixinsFromMixinSources = filterMixins(mixinsFromMixinSources, modelMixins);
+
+  mixins = mixins.concat(mixinsFromMixinSources);
```

Comment:
This may create duplicate items in the `mixins` array, when the same mixin is loaded from `mixinDirs` and `mixinSources`, is that correct? In which case you should de-duplicate the instructions.